libraries/AP_ADSB/AP_ADSB_Sagetech.cpp

```diff
@@ -0,0 +1,746 @@
+
+        } else if (now_ms - last_packet_Operating_ms >= 1000 && (
+                last_packet_Operating_ms == 0 || // send once at boot
+                // send as data changes
+                last_operating_squawk != frontend.out_state.cfg.squawk_octal ||
+                abs(last_operating_alt - frontend._my_loc.alt) > 1555 ||      // 1493cm == 49ft. The output resolution is 100ft per bit
+                last_operating_rf_select != frontend.out_state.cfg.rfSelect))
+        {
+            last_packet_Operating_ms = now_ms;
+            last_operating_squawk = frontend.out_state.cfg.squawk_octal;
+            last_operating_alt = frontend._my_loc.alt;
+            last_operating_rf_select = frontend.out_state.cfg.rfSelect;
+            send_packet(MsgTypes_XP::Operating_Set);
+
+        } else if (now_ms - last_packet_GPS_ms >= (frontend.out_state.is_flying ? 200 : 1000)) {
+            // 1Hz when not flying, 5Hz when flying
+            last_packet_GPS_ms = now_ms;
+            send_packet(MsgTypes_XP::GPS_Set);
+        }
+    }
+}
+
+void AP_ADSB_Sagetech::send_packet(const MsgTypes_XP type)
+{
+    switch (type) {
+    case MsgTypes_XP::Installation_Set:
+        send_Installation();
+        break;
+    case MsgTypes_XP::Preflight_Set:
+        send_PreFlight();
+        break;
+    case MsgTypes_XP::Operating_Set:
+        send_Operating();
+        break;
+    case MsgTypes_XP::GPS_Set:
+        send_GPS();
+        break;
+    default:
+        break;
+    }
+}
+
+void AP_ADSB_Sagetech::request_packet(const MsgTypes_XP type)
+{
+    Packet_XP pkt {};
+
+    pkt.type = MsgTypes_XP::Request;
+    pkt.id = 0;
+    pkt.payload_length = 4;
+
+    pkt.payload[0] = static_cast<uint8_t>(type);
+
+    send_msg(pkt);
+}
+
+
+void AP_ADSB_Sagetech::parse_packet_XP(const Packet_XP &msg)
+{
+#if SAGETECH_DEBUG_RX
+    gcs().send_text(MAV_SEVERITY_DEBUG, "%sRX type=%u, id=%d", frontend.GcsHeader, (unsigned)msg.type, msg.id);
+#endif
+
+    switch (msg.type) {
+    case MsgTypes_XP::ACK: {
+        // ACK received!
+        const uint8_t system_state = msg.payload[2];
+        transponder_type = (Transponder_Type)msg.payload[6];
+
+        const char* rfmode = "RF mode: ";
+        const uint8_t prev_transponder_mode = last_ack_transponder_mode;
+        last_ack_transponder_mode = (system_state >> 6) & 0x03;
+        if (prev_transponder_mode != last_ack_transponder_mode) {
+            switch (last_ack_transponder_mode) {
+            case 0: gcs().send_text(MAV_SEVERITY_INFO, "%s%sOFF",   frontend.GcsHeader, rfmode); break;
+            case 1: gcs().send_text(MAV_SEVERITY_INFO, "%s%sSTBY",  frontend.GcsHeader, rfmode); break;
+            case 2: gcs().send_text(MAV_SEVERITY_INFO, "%s%sON",    frontend.GcsHeader, rfmode); break;
+            case 3: gcs().send_text(MAV_SEVERITY_INFO, "%s%sON-ALT",frontend.GcsHeader, rfmode); break;
+            default:  break;
+            }
+        }
+
+#if SAGETECH_DEBUG_RX_ACK

```

Comment:
Again, do we really want to carry all of this debug?  

"Yes" is a valid response - both AP_Logger and GCS_MAVLink have chunks of code like this hanging around.